crates/findex/src/pyo3/py_api.rs

```diff
@@ -536,3 +518,213 @@ impl InternalFindex {
         Ok(())
     }
 }
+
+#[pyclass]
+pub struct FindexCloud(FindexCloudRust);
+
+#[pymethods]
+impl FindexCloud {
+    #[new]
+    pub fn new(token: &str, base_url: Option<String>) -> PyResult<Self> {
+        Ok(Self(pyo3_unwrap!(
+            FindexCloudRust::new(token, base_url),
+            "error reading token"
+        )))
+    }
+
+    /// Upserts the given relations between `IndexedValue` and `Keyword` into
+    /// Findex tables. After upserting, any search for a `Word` given in the
+    /// aforementioned relations will result in finding (at least) the
+    /// corresponding `IndexedValue`.
+    ///
+    /// Parameters
+    ///
+    /// - `indexed_values_and_keywords` : map of `IndexedValue` to keywords
+    /// - `master_key`                  : Findex master key
+    /// - `label`                       : label used to allow versioning
+    pub fn upsert(
+        &mut self,
+        indexed_values_and_keywords: HashMap<IndexedValuePy, Vec<&str>>,
+        master_key: &MasterKeyPy,
+        label: &LabelPy,
+    ) -> PyResult<()> {
+        let future = self.0.upsert(
+            indexed_values_and_keywords_to_rust(indexed_values_and_keywords),
+            &master_key.0,
+            &label.0,
+        );
+        // We want to forward error code returned by callbacks to the parent caller to
+        // do error management client side.
+        let rt = pyo3_unwrap!(
+            tokio::runtime::Runtime::new(),
+            "async runtime creation error"
+        );
+        pyo3_unwrap!(rt.block_on(future), "error blocking for upsert");
+        Ok(())
+    }
+
+    /// Recursively search Findex graphs for `Location` corresponding to the
+    /// given `Keyword`.
+    ///
+    /// Parameters
+    ///
+    /// - `keywords`                : keywords to search using Findex
+    /// - `master_key`              : user secret key
+    /// - `label`                   : public label used in keyword hashing
+    /// - `max_results_per_keyword` : maximum number of results to fetch per
+    ///   keyword
+    /// - `max_depth`               : maximum recursion level allowed
+    /// - `fetch_chains_batch_size` : batch size during fetch chain
+    /// - `progress_callback`       : optional callback to process intermediate
+    ///   search results.
+    ///
+    /// Returns: List[IndexedValue]
+    // use `u32::MAX` for `max_result_per_keyword`
```

Comment:
Maybe we don't need this? If you still need it, maybe you can place it above the specific line (`max_result_per_keyword = ...`)

```suggestion
```